### PR TITLE
Implement Recursive Actions (Fix #73)

### LINKED ISSUE #73: Nested Update/Effects
Have you considered this:

```javascript
app({
	update: {
		example: {
			attempt: () => alert(1)
		}
	}
})
```

It should then be possible to call `msg.example.attempt()`.

---

<details>
<summary>Meanwhile I ended up using this</summary>

```javascript
const convert = (actions, prefix = '') => Object
	.keys(actions)
	.map(name => typeof actions[name] === 'function'
		? ({[`${prefix[0].toLowerCase()}${prefix.slice(1)}${name[0].toUpperCase()}${name.slice(1)}`]: actions[name]})
		: (convert(actions[name], `${prefix}${name[0].toUpperCase()}${name.slice(1)}`))
	)
	.reduce((a, b) => Object.assign(a, b), {})

app({
	update: convert({
		example: {
			attempt: () => alert(1)
		}
	})
})
```

This maps `example.attempt` to `exampleAttempt`.

</details>

### PR DESCRIPTION
**Pull request by dodekeract**
Makes this possible:

```jsx
const { h, app } = hyperapp
/** @jsx h */

app({
	model: 0,
	reducers: {
		counter: {
			save: (_, data) => data
		}
	},
	effects: {
		counter: {
			sub: (model, actions) => actions.counter.save(model - 1)
			add: (model, actions) => actions.counter.save(model + 1)
		}
	},
	view: (model, {counter}) => (
		<div>
			<button onclick={counter.add}>
				+
			</button>
			<h1>{model}</h1>
			<button onclick={counter.sub} disabled={model <= 0}>
				-
			</button>
		</div>
	)
})
```

---

## Meta

Issue: #73

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
-
-				var effect = effects[name]
-				if (effect) {
-					return effect(model, actions, data, hooks.onError)
-				}
-
-				var update = reducers[name], _model = model
-				render(model = merge(model, update(model, data)), view, node)
-
-				hooks.onUpdate(_model, model, data)
```
Comment (dodekeract) [inline src/app.js]: Oh, I also fixed `onUpdate` not firing if `render` crashes. That's probably the time where you need `onUpdate` most.

File: src/app.js [hunk 2]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
+	function recurseActions (isReducer, pointer, actionPointer, _name) {
+		for (var key in pointer) {
+			if (!actionPointer[key]) actionPointer[key] = {}
+			var name = _name ? _name + '.' + key : key
+			if (pointer[key] instanceof Function) {
+				(function (key, name) {
+					actionPointer[key] = function (data) {
+						hooks.onAction(name, data)
+						if (isReducer) {
+							hooks.onUpdate(model, model = merge(model, pointer[key](model, data)), data)
```
Comment (dodekeract) [inline src/app.js]: `_model` is not needed anymore. It's now implicit.

File: src/app.js [hunk 3]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
+							render(model, view, node)
+						} else return pointer[key](model, actions, data, hooks.onError)
+					}
+				})(key, name)
+			} else recurseActions(isReducer, pointer[key], actionPointer[key], name)
+		}
 	}
+	recurseActions(0, effects, actions, '')
+	recurseActions(1, reducers, actions, '')
```
Comment (dodekeract) [inline src/app.js]: Could be `false/true`, but `0/1` leads to -2 bytes.

**Comment by jbucaran** [comment]
@dodekeract Interesting! 😄 

Can you clarify the objective is to be able to accomplish namespaces for actions?

**Comment by dodekeract** [comment]
@jbucaran yes.

**Comment by jbucaran** [comment]
What if you want to use namespaces for some actions, but not for others?

**Comment by dodekeract** [comment]
@jbucaran Still works, of course. Namespaces can be as deep as you want.

(tests don't fail, which proofs this claim)

**Comment by jbucaran** [comment]
@dodekeract I'm going through the code, in the meantime, could you try to think how the documentation will need to be rewritten or adapted?

**Comment by dodekeract** [comment]
@jbucaran Honestly, since this affects both reducers and effects I'd add a new topic `actions` to the Wiki. Then explain that both will be combined and can have arbitrary depth.

I'd also suggest to adapt one or more examples to use this, so that it's "advertised" to users that don't read the Wiki.

**Comment by jbucaran** [comment]
Okay, I'm done, very nice job @dodekeract! I'll ship this with the next release.

If you can help with the documentation a bit, we could ship this fater! 🤘

File: src/app.js (hunk 1 above)
Comment (jbucaran) [inline src/app.js]: @dodekeract What do you mean by "implicit"?

File: src/app.js [hunk 4]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
-
-				var update = reducers[name], _model = model
-				render(model = merge(model, update(model, data)), view, node)
-
-				hooks.onUpdate(_model, model, data)
-			}
-		}(name))
+	function recurseActions (isReducer, pointer, actionPointer, _name) {
+		for (var key in pointer) {
+			if (!actionPointer[key]) actionPointer[key] = {}
```
Comment (dodekeract) [inline src/app.js]: It creates empty objects for every key it'll later try to assign a function to. You can't just assign it to nothing.

File: src/app.js [hunk 5]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
+			var name = _name ? _name + '.' + key : key
+			if (pointer[key] instanceof Function) {
+				(function (key, name) {
+					actionPointer[key] = function (data) {
+						hooks.onAction(name, data)
+						if (isReducer) {
+							hooks.onUpdate(model, model = merge(model, pointer[key](model, data)), data)
+							render(model, view, node)
```
Comment (dodekeract) [inline src/app.js]: We can. I just wanted to keep the scope to a single change.

File: src/app.js (hunk 1 above)
Comment (dodekeract) [inline src/app.js]: Well, `_model` got replaced by moving the assignment into the `hooks.onUpdate` call.

File: src/app.js [hunk 6]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
-				render(model = merge(model, update(model, data)), view, node)
-
-				hooks.onUpdate(_model, model, data)
-			}
-		}(name))
+	function recurseActions (isReducer, pointer, actionPointer, _name) {
+		for (var key in pointer) {
+			if (!actionPointer[key]) actionPointer[key] = {}
+			var name = _name ? _name + '.' + key : key
+			if (pointer[key] instanceof Function) {
```
Comment (dodekeract) [inline src/app.js]: Not sure, but works too. Fine with changing this.

File: src/app.js [hunk 7]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
-					return effect(model, actions, data, hooks.onError)
-				}
-
-				var update = reducers[name], _model = model
-				render(model = merge(model, update(model, data)), view, node)
-
-				hooks.onUpdate(_model, model, data)
-			}
-		}(name))
+	function recurseActions (isReducer, pointer, actionPointer, _name) {
```
Comment (dodekeract) [inline src/app.js]: I'm really sure that we can't avoid that. The signature is also as good as possible AFAIK.

I agree that the signature isn't the most readable one, but try removing anything. There is nothing that isn't necessary.

File: src/app.js [hunk 8]
```diff
@@ -21,23 +21,25 @@ export default function (options) {
...
+		for (var key in pointer) {
```
Comment (dodekeract) [inline src/app.js]: Might work. I personally don't like `forEach` without arrow functions though. I'll see what the size difference is.

Comment (dodekeract) [inline src/app.js]: Oh, we actually can't get rid of the closure since name still depends on it.

File: src/app.js (hunk 4 above)
Comment (jbucaran) [inline src/app.js]: Can we get rid of the `if` and just assign it to {}?

File: src/app.js (hunk 5 above)
Comment (jbucaran) [inline src/app.js]: No worries. I'll take care of it.

**Comment by jbucaran** [comment]
@dodekeract Thanks for your comments. I'm taking care of it now.

**Comment by dodekeract** [comment]
@jbucaran Not so fast, I just saved another 6 Bytes. 🎉

File: src/app.js (hunk 4 above)
Comment (dodekeract) [inline src/app.js]: @jbucaran No, that would overwrite the actions we previously assigned to the same part of the actions tree.

**Comment by jbucaran** [comment]
Which was smaller, the one with the forEach or for loop?

**Comment by dodekeract** [comment]
@jbucaran Last commit is smallest. `forEach` enabled some further optimizations due to scoping.

**Comment by jbucaran** [comment]
👍 Thanks @dodekeract.

**Pull request closed without merging**